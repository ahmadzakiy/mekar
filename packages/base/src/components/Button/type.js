export const VARIATION = [
  'primary',
  'secondary',
]

export const SIZE = [
  'medium',
  'small',
]
