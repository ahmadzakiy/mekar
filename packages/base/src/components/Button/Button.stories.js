import Button from './Button'
import { VARIATION, SIZE } from './type'

export default {
  title: 'Components/Button',
  component: Button,
  argTypes: {
    default: {
      control: 'text',
      description: 'Slot text',
    },
    variation: {
      control: 'select',
      options: VARIATION
    },
    size: {
      control: 'select',
      options: SIZE
    },
    onClick: {
      action: 'click'
    }
  },
};

const Template = (args) => ({
  components: {
    Button
  },
  setup() {
    return { args }
  },
  template: '<Button v-bind="args">{{ args.default }}</Button>',
});

export const Default = Template.bind({});
Default.args = {
  default: 'Submit',
  variation: 'primary',
  size: 'medium'
}
