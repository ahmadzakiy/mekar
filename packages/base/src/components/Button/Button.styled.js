import styled from "vue3-styled-components";
import { color } from 'mekar-token/src/token'

const buttonProps = {
  variation: String,
  size: String,
  isDisabled: Boolean
}

const StyledButton = styled('button', buttonProps)`
  cursor: pointer;
  height: ${props => props.size === 'small' ? '36px' : '44px' };
  padding:${props => props.size === 'small' ? '0 12px' : '0 20px' };
  border: 1px solid transparent;
  text-decoration: none;
  outline: none;
  color: ${props => props.variation === 'primary' ? color.white : color.gray80 };
  font-family: sans-serif;
  font-size: 16px;
  line-height: 20px;
  background-color: ${props => props.variation === 'primary' ? color.blue60 : color.gray10 };
  transition: all .2s ease-in-out;

  ${props => props.isDisabled && `
     cursor: not-allowed;
     background-color: #f2f2f2;
     color: #b6b6b6;
  `}
`

export default StyledButton

