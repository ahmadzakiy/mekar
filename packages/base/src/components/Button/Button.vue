<template>
  <StyledButton
    :variation="variation"
    :size="size"
    :is-disabled="isDisabled"
    @click="handleClick"
  >
    <!-- @slot Use slot for text -->
    <slot />
  </StyledButton>
</template>

<script>
  import StyledButton from './Button.styled'
  import { VARIATION, SIZE } from './type'

  export default {
    name: 'Button',
    version: '1.0.0',
    components: {
      StyledButton
    },
    props: {
      /**
       * Variation of the button
       * @values primary, secondary
       */
      variation: {
        type: String,
        default: 'primary',
        validator: value => VARIATION.includes(value)
      },
      /**
       * Size of the button
       * @values medium, small
       */
      size: {
        type: String,
        default: 'medium',
        validator: value => SIZE.includes(value)
      },
      /**
       * Whether the button is disabled or not
       */
      isDisabled: {
        type: Boolean,
        default: false
      },
    },
    emits: ['click'],
    methods: {
      handleClick(event) {
        if (this.isDisabled) {
          return
        }

        /**
         * Triggers when the button clicked
         *
         * @event click
         * @type {object}
         * @arg event return event object
         */
        this.$emit('click', event)
      }
    }
  }
</script>

