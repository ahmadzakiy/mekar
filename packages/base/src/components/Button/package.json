{
  "name": "mekar-button",
  "version": "0.5.0",
  "description": "component button",
  "main": "dist/index.js",
  "scripts": {
    "clean": "rm -rf ./dist",
    "release": "npm publish"
  },
  "author": "Ahmad Zakiy <in.zakiy@gmail.com>",
  "license": "MIT",
  "devDependencies": {
    "vue": "^3.2.12"
  },
  "peerDependencies": {
    "vue": "^3.2.12"
  },
  "dependencies": {
    "vue3-styled-components": "^1.2.1"
  }
}
