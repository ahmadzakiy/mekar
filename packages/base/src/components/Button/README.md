# Mekar Button

## Installation
Install the component via NPM

```bash
yarn add mekar-button
```

## How to Use

### Import as Global Component
With this kind of import, you will include the whole bundle of the library, but you can use all components directly in Vue Instance.

```js
import { createApp } from 'vue'
import App from './App.vue'

// components
import Button from 'mekar-button'

const app = createApp(App)
app.use(Button)
app.mount('#app')
```

### Default Import
With this kind of import, you will only include component that've been imported.

```js
// components
import Button from 'mekar-button'
```

## Contribution
Everyone is welcome to contribute to Mekar Library
