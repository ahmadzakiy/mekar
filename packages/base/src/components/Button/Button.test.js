import { shallowMount } from '@vue/test-utils'
import Button from './Button'

describe('Component: Button', () => {
  let instance

  const propsObj = {
    variation: 'primary',
    size: 'medium',
    isDisabled: false,
  }

  beforeEach(() => {
    instance = shallowMount(Button, {
      slots: {
        default: 'Submit',
      },
    })
  })

  afterEach(() => {
    instance.unmount()
  })

  it('Should be named Button', () => {
    expect(instance.vm.$options.name).toBe('Button')
  })

  it('Should have class bl-button', () => {
    expect(instance.find('button').classes()).toContain('button')
  })

  it('Should match snapshot', () => {
    expect(instance.element).toMatchSnapshot()
  })

  it('Should have default props correctly', () => {
    expect(instance.props()).toEqual(propsObj)
  })

  it('Should render button text correctly', () => {
    expect(instance.find('button').text()).toBe('Submit')
  })

  it('Should handle theme correctly', async () => {
    await instance.setProps({
      variation: 'primary',
      size: 'medium',
    })

    expect(instance.vm.theme).toEqual({
      backgroundColor: '#2572b7',
      color: '#ffffff',
      height: '44px',
      padding: '0px 20px',
    })

    await instance.setProps({
      variation: 'secondary',
      size: 'small',
    })

    expect(instance.vm.theme).toEqual({
      backgroundColor: '#ebebeb',
      color: '#3d3d3d',
      height: '36px',
      padding: '0px 12px',
    })
  })

  it('Should render disabled state correctly', async () => {
    await instance.setProps({
      isDisabled: true,
    })
    expect(instance.classes()).toContain('is-disabled')

    await instance.trigger('click')
    expect(instance.emitted()).not.toHaveProperty('click')
  })

  it('Should emit click correctly', async () => {
    await instance.trigger('click')
    expect(instance.emitted()).toHaveProperty('click')
  })
})
