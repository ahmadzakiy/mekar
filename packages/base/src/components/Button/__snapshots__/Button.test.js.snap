// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Button Should match snapshot 1`] = `
<button
  class="edSiXT"
>
  <!-- @slot Use slot for text -->
  
  Submit
  
</button>
`;
